File: src/lib/readerPrefs.ts
/**
 * readerPrefs.ts
 * ───────────────────────────────────────────────────────────────────────────
 * Single source of truth for the reader's per-device preferences:
 * study level, audience, translation. This module owns:
 *
 *   • the localStorage key names every client persists under
 *   • the default values for first-time visitors
 *   • the set of valid values (and their type-guard predicates)
 *   • the public-domain translation catalog
 *
 * Every web React context (StudyLevelContext, AudienceContext,
 * TranslationContext) reads from this module. The iOS app mirrors these
 * keys + defaults in `ios/LearnOfChrist/Models/ReaderPrefs.swift` (to
 * be added — see `docs/IOS-PORT-AUDIT-2026-05-12.md` punch-list item
 * #10) so a future `user_reading_prefs` sync row is byte-compatible
 * across web and native.
 *
 * IMPORTANT — DO NOT rename a `STORAGE_KEY` once it's shipped. Users
 * have values persisted under these strings; renaming would silently
 * reset their preferences. Add a new key + migration path if the schema
 * needs to evolve.
 */

import type { StudyLevel as RichStudyLevel } from '@/data/study-chapters/types';

/* ─── Study level ────────────────────────────────────────────────────── */

/** Re-exported so consumers can import the type and storage contract from
 *  one place. The canonical type definition lives in
 *  `src/data/study-chapters/types.ts` (and `lib/StudyLevelContext.tsx`).
 */
export type StudyLevel = RichStudyLevel;

export const STUDY_LEVEL_STORAGE_KEY = 'loc-study-level';

/** Default for new readers. Beginner risks training people that the site
 *  is shallow; Deep Dive risks a wall-of-text first impression. */
export const DEFAULT_STUDY_LEVEL: StudyLevel = 'intermediate';

export const STUDY_LEVEL_VALUES: readonly StudyLevel[] = [
  'beginner',
  'intermediate',
  'deep',
] as const;

export function isStudyLevel(v: unknown): v is StudyLevel {
  return v === 'beginner' || v === 'intermediate' || v === 'deep';
}

/* ─── Audience ───────────────────────────────────────────────────────── */

import type { Audience as RichAudience } from '@/data/study-chapters/types';

export type Audience = RichAudience;

export const AUDIENCE_STORAGE_KEY = 'loc-audience';

/** Set to '1' when the active user's account flag forces audience to
 *  'kids' (e.g. a parent-configured child account). When this key is '1',
 *  the picker hides itself and audience is locked to 'kids'. */
export const AUDIENCE_LOCK_STORAGE_KEY = 'loc-audience-lock';

/** Default for new readers. Most visitors are adults; teens and
 *  parents-with-kids will pick their audience explicitly. */
export const DEFAULT_AUDIENCE: Audience = 'adults';

export const AUDIENCE_VALUES: readonly Audience[] = [
  'adults',
  'youth',
  'kids',
] as const;

export function isAudience(v: unknown): v is Audience {
  return v === 'adults' || v === 'youth' || v === 'kids';
}

/* ─── Translation ────────────────────────────────────────────────────── */

/** Bible translation chosen for verse rendering. Stored as the
 *  translation's `abbreviation` (lowercase). */
export type TranslationAbbr = string;

export const TRANSLATION_STORAGE_KEY = 'loc-translation';

export const DEFAULT_TRANSLATION: TranslationAbbr = 'kjv';

/** Public-domain translation catalog. KJV stays at the top as the
 *  default; the rest are alphabetised by full name so the dropdown is
 *  easy to scan.
 *
 *  Licensed translations (NIV/ESV/CSB/etc.) are NOT in v1 — see
 *  `docs/MOBILE-PLATFORM-STRATEGY.md → "Out of scope for v1"`. When
 *  they arrive, they'll be served by the `bible-translation-proxy`
 *  Edge Function and slotted in here (kept separate from the
 *  public-domain list so offline clients can render
 *  ahead-of-license-check).
 */
const KJV = { abbreviation: 'kjv', name: 'King James Version' };
const PUBLIC_DOMAIN_REST = [
  { abbreviation: 'asv', name: 'American Standard Version' },
  { abbreviation: 'bsb', name: 'Berean Standard Bible' },
  { abbreviation: 'dra', name: 'Douay-Rheims American' },
  { abbreviation: 'jst', name: 'Joseph Smith Translation' },
  { abbreviation: 'lsv', name: 'Literal Standard Version' },
  { abbreviation: 'nwt', name: 'New World Translation' },
  { abbreviation: 'web', name: 'World English Bible' },
].sort((a, b) => a.name.localeCompare(b.name));

export const PUBLIC_DOMAIN_TRANSLATIONS: readonly {
  abbreviation: string;
  name: string;
}[] = [KJV, ...PUBLIC_DOMAIN_REST];

export function isKnownTranslation(abbr: unknown): abbr is TranslationAbbr {
  if (typeof abbr !== 'string') return false;
  return PUBLIC_DOMAIN_TRANSLATIONS.some((t) => t.abbreviation === abbr);
}

/* ─── Cross-platform sync row ────────────────────────────────────────── */

/** Shape the `user_reading_prefs` row uses (and the future
 *  `v1_sync_user_data` payload). Keep additive — clients on older
 *  versions must tolerate unknown keys.
 *
 *  Mirrored byte-for-byte by `ReaderPrefs` on iOS / Android so the same
 *  serialized JSON round-trips through every client.
 */
export interface ReaderPrefsSyncRow {
  study_level: StudyLevel;
  audience: Audience;
  /** True iff the user's account is locked to the kids audience. */
  audience_lock: boolean;
  translation: TranslationAbbr;
}

/** Build a sync row from the in-memory values. Mostly here so tests
 *  (and future RPC callers) have a single function to call. */
export function makeReaderPrefsSyncRow(input: {
  studyLevel: StudyLevel;
  audience: Audience;
  audienceLock: boolean;
  translation: TranslationAbbr;
}): ReaderPrefsSyncRow {
  return {
    study_level: input.studyLevel,
    audience: input.audience,
    audience_lock: input.audienceLock,
    translation: input.translation,
  };
}

File: ios/LearnOfChrist/Models/ReaderPrefs.swift
// ReaderPrefs.swift
// ────────────────────────────────────────────────────────────────────────────
// Swift twin of `src/lib/readerPrefs.ts`. Same storage keys, same default
// values, same valid value sets — that's what makes the depth + audience
// + translation preferences round-trip across web and iOS without losing
// fidelity.
//
// Persistence: stored in UserDefaults via @AppStorage with the keys in
// `ReaderPrefs.Key`. When a user signs in for the first time, the apps
// (and web) will push these into the `user_reading_prefs` Supabase row
// (migration 060/061) using the column shape captured in
// `ReaderPrefsSyncRow` below. Until then, prefs are per-device.
//
// IMPORTANT — DO NOT rename the Key string values once they're shipped.
// Users have values persisted under these strings; renaming would
// silently reset their preferences. Match the TypeScript module
// byte-for-byte.

import Foundation

enum ReaderPrefs {

    // MARK: - Storage keys

    /// @AppStorage keys for the reader-prefs trio. Match
    /// `src/lib/readerPrefs.ts` exactly.
    enum Key {
        /// Bible study depth. Values: `beginner` | `intermediate` | `deep`.
        static let studyLevel    = "loc-study-level"
        /// Reader audience. Values: `adults` | `youth` | `kids`. iOS v1
        /// picker only exposes adults + youth; kids is web-only until
        /// the kids decoder + registry exist on iOS (audit punch-list
        /// open item, ~3-day add).
        static let audience      = "loc-audience"
        /// `"1"` when the active account is locked to kids audience
        /// (child-account profile). Picker hides itself when set.
        static let audienceLock  = "loc-audience-lock"
        /// Translation abbreviation (lowercase). One of the
        /// `publicDomainTranslations` below in v1.
        static let translation   = "loc-translation"
    }

    // MARK: - Defaults

    /// Default for new readers. Beginner risks training people that the
    /// site is shallow; Deep risks a wall-of-text first impression.
    static let defaultStudyLevel: StudyLevel = .intermediate

    /// Most visitors are adults; teens and parents-with-kids pick
    /// explicitly.
    static let defaultAudience: Audience = .adults

    /// KJV by default. Public domain, broadly recognised, doesn't tie
    /// us to a licensed translation.
    static let defaultTranslation: TranslationAbbr = "kjv"

    // MARK: - Translation catalog

    typealias TranslationAbbr = String

    /// One translation entry. Matches the web shape
    /// `{ abbreviation, name }`.
    struct Translation: Sendable, Hashable, Identifiable {
        let abbreviation: String
        let name: String
        var id: String { abbreviation }
    }

    /// Public-domain translations only. KJV at the top as the default;
    /// the rest alphabetised by full name so the picker is easy to scan.
    /// Mirrors `PUBLIC_DOMAIN_TRANSLATIONS` in `src/lib/readerPrefs.ts`.
    ///
    /// Licensed translations (NIV / ESV / CSB / etc.) are explicitly
    /// out of scope for v1 — see
    /// `docs/MOBILE-PLATFORM-STRATEGY.md → "Out of scope for v1"`.
    static let publicDomainTranslations: [Translation] = {
        let kjv = Translation(abbreviation: "kjv", name: "King James Version")
        let rest: [Translation] = [
            Translation(abbreviation: "asv", name: "American Standard Version"),
            Translation(abbreviation: "bsb", name: "Berean Standard Bible"),
            Translation(abbreviation: "dra", name: "Douay-Rheims American"),
            Translation(abbreviation: "jst", name: "Joseph Smith Translation"),
            Translation(abbreviation: "lsv", name: "Literal Standard Version"),
            Translation(abbreviation: "nwt", name: "New World Translation"),
            Translation(abbreviation: "web", name: "World English Bible"),
        ].sorted { $0.name < $1.name }
        return [kjv] + rest
    }()

    /// Look up a translation by its abbreviation, falling back to the
    /// default when the stored value is unknown (e.g. a user signed in
    /// on web with a translation we removed).
    static func translation(forAbbreviation abbr: String) -> Translation {
        publicDomainTranslations.first { $0.abbreviation == abbr }
            ?? publicDomainTranslations.first { $0.abbreviation == defaultTranslation }!
    }

    static func isKnownTranslation(_ abbr: String) -> Bool {
        publicDomainTranslations.contains { $0.abbreviation == abbr }
    }

    // MARK: - Sync row (future user_reading_prefs payload)

    /// Shape of the `user_reading_prefs` row (migrations 060 + 061) and
    /// the `claim_local_data` payload key. Additive going forward — older
    /// clients must tolerate unknown keys, never alter existing keys.
    ///
    /// Mirrors `ReaderPrefsSyncRow` in `src/lib/readerPrefs.ts`.
    struct SyncRow: Codable, Sendable, Hashable {
        let study_level: String
        let audience: String
        let audience_lock: Bool
        let translation: String
    }

    /// Build a sync row from the in-memory values. Mostly here so the
    /// future SyncService and any tests have a single function to call.
    static func makeSyncRow(
        studyLevel: StudyLevel,
        audience: Audience,
        audienceLock: Bool,
        translation: TranslationAbbr
    ) -> SyncRow {
        SyncRow(
            study_level: studyLevel.rawValue,
            audience: audience.rawValue,
            audience_lock: audienceLock,
            translation: translation
        )
    }
}

// MARK: - UI metadata extensions

extension StudyLevel {
    /// Display name in the picker. Mirrors web `STUDY_LEVELS[*].label`.
    /// Names describe content scope, not the reader's skill.
    var label: String {
        switch self {
        case .beginner:     return "Quick"
        case .intermediate: return "Standard"
        case .deep:         return "Deep"
        }
    }

    /// Short description shown under the row in the full picker.
    var blurb: String {
        switch self {
        case .beginner:
            return "Scripture, big idea, and one takeaway."
        case .intermediate:
            return "Adds commentary, themes, and connections."
        case .deep:
            return "Full study with original languages and cross-references."
        }
    }

    /// SF Symbol for the picker chip.
    var symbol: String {
        switch self {
        case .beginner:     return "1.circle"
        case .intermediate: return "2.circle"
        case .deep:         return "3.circle"
        }
    }
}

extension Audience {
    var label: String {
        switch self {
        case .adults: return "Adults"
        case .youth:  return "Youth"
        case .kids:   return "Kids"
        }
    }

    var blurb: String {
        switch self {
        case .adults:
            return "Full study, all topics."
        case .youth:
            return "Same study, retuned for teens."
        case .kids:
            return "Story, big idea, family talk-together."
        }
    }

    var symbol: String {
        switch self {
        case .adults: return "person"
        case .youth:  return "figure.stand"
        case .kids:   return "figure.and.child.holdinghands"
        }
    }

    /// Audiences the iOS v1 reader actually supports. Kids is web-only
    /// for now — the iOS kids decoder + registry aren't built (see audit
    /// punch-list open item). When kids ships, return all three.
    static var iOSSupported: [Audience] { [.adults, .youth] }
}
